### PR TITLE
Add aircraft bias converter

### LINKED ISSUE #804: Create IODA Converter for Aircraft Bias Correction Files
## Description
Following the examples from satellite bias correction, develop converter to convert GSI aircraft bias correction files from GSI format to IODA format

## Requirements

Code to create aircraft bias correction coefficient files for input to aircraft bias correction testing in UFO.  The initial file to be produced would be for the 2020121500 case so that it can be tested as an extension to [existing test cases in UFO](https://github.com/JCSDA-internal/ufo/tree/develop/test/testinput/instrumentTests/Aircraft_obs).

## Acceptance Criteria (Definition of Done)
Create BC files that contain the same information as those in the GSI.

## Dependencies
No dependencies

Required Steps:

- [x] Obtain aircraft bias correction files from operational GSI for 2020121500 (gdas.t00z.abias_air)
- [ ] Using the [exisiting satbias converters ](https://github.com/JCSDA-internal/ioda-converters/tree/develop/src/satbias)as a template, create a converter to for aircraft BC files.  This will require C++ coding!  Example satbias yamls are [here.](https://github.com/JCSDA-internal/ioda-converters/blob/develop/test/testinput/satbias_converter_amsua.yaml)
- [ ] Using ncdump, compare the created IODA bias files with the input files.

For reference, these files are read in to GSI in [aircraftinfo.f90](https://github.com/NOAA-EMC/GSI/blob/master/src/gsi/aircraftinfo.f90).

### PR DESCRIPTION
**Pull request by kevindougherty-noaa**
## Description

The following scripts compile successfully and create an output aircraft bias .nc4 file with information about the existing predictors and tail ids. 

## Issue(s) addressed

Resolves #804

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
Thanks @kevindougherty-noaa this is a good start, can you start to look at the changes I have here: https://github.com/JCSDA-internal/ioda-converters/compare/feature/refactor_satbias and use the new file format structure? I can help with doing this if needed.

Also check the coding norms!

File: test/testinput/acftbias_converter.yaml
Comment (CoryMartin-NOAA) [inline test/testinput/acftbias_converter.yaml:3]: So this list should only be 3 instead of 9, 4-6, 7-9 should get appended from 1-3 in the code somehow.

File: src/acftbias/GsiAircraftBiasReader.h
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasReader.h:2]: Change all these copyrights to 2023 NOAA NWS NCEP EMC

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:45]: "records"

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:95]: this should be 3, 3 predictors, each has coeffs, ob errors and background errors, right?

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:100]: no satinfo file here :-)

File: test/testinput/acftbias_converter.yaml
Comment (CoryMartin-NOAA) [reply test/testinput/acftbias_converter.yaml:3]: Probably need to ask @ADCollard about what these errors should be saved as.

Comment (CoryMartin-NOAA) [reply test/testinput/acftbias_converter.yaml:3]: the background error ones should be stored as BiasCoeffErrors, not sure about the obs errors...

**Comment by kevindougherty-noaa** [comment]
Newest commit changes for the converter compiles but when testing, it throws a segmentation fault with no other indication of why.

File: src/acftbias/CMakeLists.txt
Comment (CoryMartin-NOAA) [inline src/acftbias/CMakeLists.txt:1]: 2023 NOAA NWS NCEP EMC

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:82]:
```suggestion
                                                "biasCoeffErrors/"+predictors[i],
```

Comment (CoryMartin-NOAA) [reply src/acftbias/GsiAircraftBiasConverter.cc:82]: Note this will require a re-copy of the testoutput file

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:73]: I think we might just want `ObError` here but want others to chime in here @ADCollard @shlyaeva @BenjaminRuston @gmao-yzhu @huishao-r @gmao-msienkie

File: src/acftbias/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasConverter.cc:113]:
```suggestion
          std::to_string(gsi_npredictors) + " (same as number of predictors in GSI aircraft bias file)";
```

File: src/acftbias/GsiAircraftBiasReader.cc
Comment (CoryMartin-NOAA) [inline src/acftbias/GsiAircraftBiasReader.cc:66]: Still need to save datetime as something like "last_cycle_updated"

**Comment by CoryMartin-NOAA** [comment]
This is getting close, I think the remainder is just some tidying up, marking as ready for review for discussion. Main question remaining is what to do with the datetimes. Should those be IODA datetime objects or integer values, etc.

**Comment by CoryMartin-NOAA** [comment]
Note the base branch is not the sprint branch but my satbias refactoring branch, this is just to see the changes made here and not both. This branch should be merged in after that in #1360

**Review by BenjaminRuston — APPROVED** [review]
this is working for me with the associated ctest and I also tested it with a file from @ADCollard from a while ago `gdas.t00z.abias_air`

ideally would test end-to-end but think should be good to move forward in sprint branches

File: src/varbc_gsi/generate_aircraft_bias_csv.py
Comment (huishao-r) [inline src/varbc_gsi/generate_aircraft_bias_csv.py:1]: I actually think acftbias/ seems consistent with current repo (which has satbias/). Is there any reason to change the dir name to varbc_gir?

Comment (CoryMartin-NOAA) [reply src/varbc_gsi/generate_aircraft_bias_csv.py:1]: two reasons in my mind, but happy to discuss. 1) these are GSI VarBC converters 2) it consolidates 2 directories of ~4 files into 1 directory with ~8 files, just seemed cleaner.

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (shlyaeva) [inline src/varbc_gsi/GsiAircraftBiasConverter.cc]: Seems to be some unintended ascii art (probably from auto-commenting by editor?)

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (shlyaeva) [inline src/varbc_gsi/GsiAircraftBiasConverter.cc]: For my understanding: what does the "predictor observation error" mean?

**Review by shlyaeva — COMMENTED** [review]
I didn't test, but I reviewed the code, looks good to me (nitpick comment and one question below)

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: Good question. @ADCollard any comment? Only the value for constant is even populated. See in the linked issue: "observation error sums for each predictor (currently only column 5 is populated)"

Comment (gmao-msienkie) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: I believe that column (sixth column in the aircraft bias coefficient file) has the number of observations for that tail number in the cycle, that would have contributed to the update of the coefficient.

Comment (CoryMartin-NOAA) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: @gmao-msienkie hmm, that would actually make a lot more sense. What are the next two columns supposed to be then? I know the last 3 are the background errors.

Comment (BenjaminRuston) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: this naming of variable and comments is a bit odd.  It really is the predictor error the `Ob` in all these names is just to let us know it's Obs related.
`obSubVarError`,,,, `obBiasCoeffErrors` makes a bit more sense as these are again predictor or coefficient error not `ObErrors`

and lastly are you saying just for the aircraft bias coefficient file the predictor (aircraft ascent rate) does not have any error assigned?

Comment (gmao-msienkie) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: I think that you would need to check with @gmao-yzhu to be sure of the intent (or original intent) of the three columns here.  She had worked on a couple of different methods for the bias correction - this one with ascent and ascent-squared, and another one that calculated coefficients to be applied depending on phase of flight (ascending or descending) so there would have been counts for each type (level flight, ascending, or descending).  Eventually only the method with vvel and vvel-squared ended up being used in NCEP operations, so the other columns are not needed.  Our code at GMAO still has the reference to the 'aircraft_t_bc_pof' method as well as my 'aircraft_bc_ext' that was used in our MERRA2 analysis.

Comment (ADCollard) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: Just chiming in to say that @gmao-msienkie 's explanation is correct as far as I know.   The zeroed columns are used when the three phases of flight are treated differently.  I do not have a strong opinion on whether to retain this capability.

My explanation here [here](https://github.com/JCSDA-internal/ioda-converters/issues/804#issuecomment-1212274832) was not entirely correct (and confusing), so I will amend that.

File: test/testinput/acftbias_converter.yaml
Comment (CoryMartin-NOAA) [inline test/testinput/acftbias_converter.yaml:4]: per the convo here: https://github.com/JCSDA-internal/ufo/pull/2979#discussion_r1324862968 does this need to be `windUpward`?

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (gmao-yzhu) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: @gmao-msienkie @ADCollard Your explanation is correct. To keep the number of data count columns and the number of predictors the same may allow us to have a little more flexibility in the future. 
@gmao-msienkie Could you please make sure that changes made in this PR can be applied to your 'aircraft_bc_ext' that was used in our MERRA2 analysis? 

@BenjaminRuston I also feel that obBiasCoeffErrors makes more sense.

Comment (CoryMartin-NOAA) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: Just so I understand what you are proposing:
you want something like:
`numObsAssimilated/constant(nrecs, nvars)`
`numObsAssimilated/vertical_velocity(nrecs, nvars)`
`numObsAssimilated/vertical_velocity_squared(nrecs, nvars)`
?

File: test/testinput/acftbias_converter.yaml
Comment (gmao-yzhu) [reply test/testinput/acftbias_converter.yaml:4]: "windUpward" is not correct, ascentRate is not precise, as it may be descentRate, it is aircraft ascent/descent rate.

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (gmao-yzhu) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: Is this why you are discussing about the number of data count columns?  
When you are using "constant" + "vertical_velocity" + "order" in yaml file, you consider the predictor number to be 3 or 2?

File: test/testinput/acftbias_converter.yaml
Comment (CoryMartin-NOAA) [reply test/testinput/acftbias_converter.yaml:4]: Well @BrettHoover-NOAA 's point was the only thing in the IODA conventions is `windUpward`, so we as a group need to figure out the proper name. As I agree, wind is not the right answer here, because it is more like `aircraftVerticalVelocity`. See spreadsheet: https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/conventions/tbls.html

Comment (gmao-yzhu) [reply test/testinput/acftbias_converter.yaml:4]: I saw "aircraftVelocity" in the spreadsheet, so I'm fine with aircraftVerticalVelocity.

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (CoryMartin-NOAA) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: For what it's worth, unless you have a strong preference, I'd rather not have different number of obs assimilated for each predictor. I would advocate for:
`numObsAssimilated(nrecs, nvars)`
where in the aircraft case, `nrecs` = ntailnumbers, and `nvars`=1=`airTemperature` (for now)
and in the radiance case, `nrecs`=1 and `nvars`=`nchans`
because I don't particularly want to clutter the radiance files with a bunch of new groups, and we would need to be consistent across both radiance and aircraft VarBC files.

File: test/testinput/acftbias_converter.yaml
Comment (CoryMartin-NOAA) [reply test/testinput/acftbias_converter.yaml:4]: and then `aircraftVerticalVelocity_order_2`?

File: src/varbc_gsi/GsiAircraftBiasConverter.cc
Comment (gmao-yzhu) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: If consistency is your priority, keeping only one slot for observation number is fine with me. 

Since you mentioned radiance bias correction, for all-sky radiance, I would strongly suggest to use the observation number that are actually used in the radiance bias correction calculation, not the observation number that are assimilated.

Comment (CoryMartin-NOAA) [reply src/varbc_gsi/GsiAircraftBiasConverter.cc]: Good point, perhaps it should be something like `numObsUsed`?

File: test/testinput/acftbias_converter.yaml
Comment (BrettHoover-NOAA) [reply test/testinput/acftbias_converter.yaml:4]: @CoryMartin-NOAA The predictor function automatically appends a `"_order_N" to the variable names on my side, the names in the ctest end up being pretty long: `predictor_obs_metadata_predictor_order_1_aircraftVerticalVelocity_airTemperature`, in this case.

I have code changes I can deploy that swap out `windUpward` with `aircraftVerticalVelocity`, if you want me to push them. ctest should pass.

Comment (CoryMartin-NOAA) [reply test/testinput/acftbias_converter.yaml:4]: I'm okay with that @BrettHoover-NOAA but I would like to confirm that others are fine with the name `aircraftVerticalVelocity` before we do a further commit/update test files.

**Comment by BenjaminRuston** [comment]
great thanks everyone for the input this looks good to go correct, will likely wait for the morning but I'm doubling down on approval this seems ready

**Review by CoryMartin-NOAA — CHANGES_REQUESTED** [review]
A couple of minor things based on discussion that I want to update before this is approved/merged

**Review by CoryMartin-NOAA — APPROVED** [review]

**Review by shlyaeva — APPROVED** [review]

**Review by huishao-r — APPROVED** [review]

**Comment by huishao-r** [comment]
@kevindougherty-noaa 
It shows there are conflicts that must be resolved. Would you please take a look:
Conflicting files
src/varbc_gsi/SatBiasConverter.cpp
src/varbc_gsi/satbias_upgrader.py

**Comment by BenjaminRuston** [comment]
@CoryMartin-NOAA I'm going ahead to test this again, and then will merge into your branch

also that branch also seems ready to go

**Comment by BenjaminRuston** [comment]
an `ncdump` of a resulting file produces this for `nvars`:
```
 nvars = 0 ;
```

**Comment by BenjaminRuston** [comment]
in the bias coefficient section do we need to differentiate the `variables` for aircraft even though it is just a single variable of `temperature`

```
aircraftVerticalVelocity
```

**Comment by CoryMartin-NOAA** [comment]
@BenjaminRuston we probably would want:
`variables(nvars) = [airTemperature`
where nvars=1

In your above example, is that `nvars` the dimension or the variable that is 0?

**Pull request merged**
